### ISSUE TITLE
Installation issue: Can't change LLVM version as dependency in Spack environment

### ISSUE THREAD

**Issue by EthanS94** [LLNL/spack#29272]
### Steps to reproduce the issue

```console
$ # using environment with paraview@master and trying to set versions for llvm and llvmlib
$ cat spack.yaml
# This is a Spack Environment file.
#
# It describes a set of packages to be installed, along with
# configuration settings.
spack:
  # add package specs to the `specs` list
  specs: [mesa]
  view: true
  packages:
    llvm:
      version: [7]
    libllvm:
      version: [7]
    llvm-amdgpu:
      version: [7]
$ spack concretize -f
...
 -   lqsnhre          ^llvm@12.0.1%gcc@7.5.0+clang~code_signing
```

### Error message

N/A

### Information on your system

* **Spack:** 0.17.1-1378-8cefe38b0e
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-zen2
* **Concretizer:** clingo

### Additional information

@haampie @trws

[spack.yaml.txt](https://github.com/spack/spack/files/8164260/spack.yaml.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by EthanS94** [comment]
@kwryankrattiger Is this related to the llvmlib virtual package?

**Comment by kwryankrattiger** [comment]
Off hand I don't think this would be caused by virtual packages, but the concretizer works in mysterious ways. It may be that the concretizer is ignoring your request for version 7, I have also seen some strange behavior around trying to constrain versions using the packages.yaml and getting unexpected results.

Does it work if you add `llvm@7` to the environment and concretize together?

The thing that is immediately strange to me here is you are specifying version constraints for `llvm`, `libllvm`, and `llvm-amdgpu`. The `llvm` one is fine, `libllvm` is not needed, and `llvm-amdgpu` is not valid since there is now version 7, the latest is 5.0.0 and corresponds _loosly_ to llvm 14.0.0. But I don't think that really matters here.

**Comment by EthanS94** [comment]
> The thing that is immediately strange to me here is you are specifying version constraints for llvm, libllvm, and llvm-amdgpu

Yeah. I just did that to see if I could get anything to change the version. The main thing is I've been using this for a while:

```yaml
packages:
  llvm:
    version: ['7.1.0']
```

And it has stopped working recently. I don't know why. I'll try adding `llvm@7` to the environment.

**Comment by EthanS94** [comment]
Adding `llvm@7` to the environment:

```yaml
  specs:
  - llvm@7
```

This resolved to `  -   lme2szl          ^llvm@7%intel@19.0.4.243~clang ...` in the spec which failed to fetch when building.

```console
==> Adding package llvm@7 to mirror
==> Cannot find version 7 in url_list
==> Cannot find version 7 in url_list
==> Cannot find version 7 in url_list
==> Warning: Error while fetching llvm@7
  All fetchers failed for spack-stage-llvm-7-lme2szlccod6nnnsocwdhzamwxfwz73c
```

Explicitly adding `llvm@7.1.0` seems like it's going ok. I could use this as a workaround

**Comment by EthanS94** [comment]
@kwryankrattiger I've narrowed it down to https://github.com/spack/spack/commit/991438a242423b9abba726874d974626f11fbd05 . If I use spack at `991438a242423b9abba726874d974626f11fbd05~1`, I do get the correct version in the concretized spec. Any later and I hit this issue.

I'm guessing this is the same issue as https://github.com/spack/spack/issues/7950
